### PR TITLE
Add role observer

### PR DESCRIPTION
**Pull request by BusyJay**
`RoleObserver` is used to receive notification when role of a peer changes.

### CODE REVIEW

File: src/raftstore/coprocessor/dispatcher.rs [hunk 1]
```diff
@@ -63,6 +65,42 @@ impl Registry {
     pub fn register_split_check_observer(&mut self, priority: u32, sco: BoxSplitCheckObserver) {
         push!(priority, sco, self.split_check_observers);
     }
+
+    pub fn register_role_observer(&mut self, priority: u32, ro: BoxRoleObserver) {
+        push!(priority, ro, self.role_observers);
+    }
+}
+
+macro_rules! try_loop_ob {
```
Comment (overvenus) [inline src/raftstore/coprocessor/dispatcher.rs]: Better add some comments for macros.

File: src/raftstore/store/worker/apply.rs [hunk 2]
```diff
@@ -167,10 +167,33 @@ pub enum ExecResult {
     DeleteRange { ranges: Vec<Range> },
 }
 
+struct ApplyCallBack {
```
Comment (overvenus) [inline src/raftstore/store/worker/apply.rs]: Is it an unrelated change to `role observer`?

**Comment by BusyJay** [comment]
PTAL

File: src/raftstore/coprocessor/dispatcher.rs [hunk 3]
```diff
@@ -63,6 +65,54 @@ impl Registry {
...
+macro_rules! loop_ob {
+    // Execute a hook, return early if error is found.
+    (_exec _res, $o:expr, $hook:ident, $ctx:expr, $($args:tt)*) => {
+        $o.$hook($ctx, $($args)*)?
+    };
+    // Execute a hook.
+    (_exec _tup, $o:expr, $hook:ident, $ctx:expr, $($args:tt)*) => {
+        $o.$hook($ctx, $($args)*)
+    };
+    // When the try loop finish successfully, the value to be returned.
```
Comment (siddontang) [inline src/raftstore/coprocessor/dispatcher.rs]: finishes

**Comment by siddontang** [comment]
Rest LGTM
